### ISSUE TITLE
Windows thread support

### ISSUE THREAD

**Issue by brano543** [libspatialindex/libspatialindex#59]
Hello, can you add thread support for the windows? Now its only possible with the POSIX threads https://www.sourceware.org/pthreads-win32/. I want to use the libspatialindex as database and those semaphores can help me to keep the consistency of data stored when there will be many queries. I am not an expert on this so correct me if i am wrong about semaphores that with them  i can use that library as a database,because they provide consistency of the data? You know,there will be many requests coming in and i think that these locks will prevent the data from being corrupted. I just heard about this at school, we didn´t go into much details :)

**Comment by davisp** [comment]
I wonder if it wouldn't be a better idea to remove concurrency control completely from libspatialindex and instead document that users need to manage it. I'm not convinced its correct as currently implemented and it seems like something that is rather orthogonal to most of the rest of the project.

**Comment by hobu** [comment]
libspatialindex is not a database. Using as one is a recipe for pain and sorry.

There is no plan to implement windows thread support at this time. We'd happily take a patch, or a patch that is aligned with the C++11 efforts that are ongoing by @normanb and @davisp and myself.

**Comment by brano543** [comment]
I still think i have a quite misunderstanding how databases work. I have read the code of your library, at least the parts i am using - StorageManager and Rtree and i didn´t think there was something that was missing. I bulk load data into the tree and then i need to serve a lot of requests. For every document in MongoDB i need to do spatial query on r-tree and thanks to those locks i can do it or am i wrong? This sounds to me as a database.

**Comment by normanb** [comment]
@davisp and myself currently use libspatialindex (and others) as a secondary index to Cloudant. It isn't a primary index for our database. Even then, we have had some issues and @davisp added a transactional storage container (which we will in the in the future try to put in the core libspatialindex code base). We benefit from some erlang features such as process mgmt which I had originally put as a native interface (NIF) but are much more secure in a process. I agree with @davisp about removing the threading from the library. @hobu @davisp and I are actively working on moving libspatialindex to a C++11 codebase.

**Comment by kounoupis** [comment]
It does not make sense to have threading support inside the library.

It used to be the case that the library had fine grained locking to support
concurrent readers and writers, but that ended up being slower than locking
the whole tree, in most scenarios, hence we got rid of them and just
settled for locking the root.

It is better if the user takes care of it at a higher level.

Regarding the definition of a database, usually people think of relational
databases when they talk about databases, and this library only implements
indices, which is a tiny part of a database system (that offer their own
sophisticated storage managers, query languages, indices, query optimizers,
etc.).

But, in any case, as long as it does what you need it do and it works, you
can call it whatever you want :)

**Comment by brano543** [comment]
@normanb Yes,i would like it to behave like secondary index to MongoDB, but i am not familiar with their code and that would take me so much time i don´t have right now. So i was thinking i will use libspatialindex to get IDs of objects i will actually look up in MongoDB.
@kounoupis  Okay, i have it compiled with those locks, but i won´t count on that so much. You said i need to take care of it at higher level, you meant the distribution layer? I was planning to use asynchronous Router-Dealer pattern as describer here http://zguide.zeromq.org/page:all#Shared-Queue-DEALER-and-ROUTER-sockets and example here https://github.com/imatix/zguide/blob/master/examples/C%2B%2B/asyncsrv.cpp. However this only solves that many clients can connect to Router and Router waits for answers from Dealer. But since many workers can call lets say function RangeQuery or Insert, if there wasn´t a line Tools::LockGuard lock(&m_lock); which locks the method with semaphore until its finished, it wouldn´t work or am i wrong? I am not sure if the OS in my case Windows cares of this stuff or it needs to be explicitly said in code like you said it. Please clarify. Anyways i am gonna try it with and without that line, i just wanted to know how its supposed to behave, it is really hard to debug things like that :)

**Comment by kounoupis** [comment]
I was making a general comment, mainly addressing the developers of the
library, agreeing with the fact that, in the future, the library should not
handle locking (as it currently does).

For your particular situation, here is how locking in this library works:
1. If there is a writer, the first writer locks the root of the tree and
has exclusive access to the whole tree. All readers block.
2. If there is one or more readers, all readers to the tree are granted
read access, and any subsequent writers block, until all readers are done.
3. I don't remember how starvation (of writers) is handled, or if it's
handled at all. I bet that it is not.

So, if you compile the library with locking support, there is no set of
actions that can corrupt the index, and hence you don't have to worry about
it.

Having said that, my personal recommendation is to forgo the locking
capabilities of the spatial index library, and figure out a way to
accomplish what I described in #1, #2, and #3 above, in some higher layer
in your code. I cannot recommend any specific way of doing that, but there
are plenty of other libraries out there that can accomplish these goals.

**Comment by hobu** [comment]
:+1:  to removing thread support in libspatialindex as it is currently implemented. I have pushed a branch to the C++11 effort #60 to do just that.
